feat: Add work email validation for SSO authentication

yashkrishan opened the pull request:
- Block new users with generic email providers (Gmail, Yahoo, Outlook, etc.)
- Block all new GitHub signups (existing users can still sign in)
- Allow legacy users with generic emails to continue using the service
- Add server-side validation in /sso/login and /signup endpoints
- Add email domain validation utilities (extract_domain, is_generic_email)
- Return 403 Forbidden with clear error messages for blocked attempts

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Signups via Google SSO using common personal/generic email domains are now blocked for new accounts (existing legacy accounts with such emails remain allowed).
  * New GitHub-based signups are blocked unless the GitHub provider is already linked to an existing account.
  * Blocked registration attempts return clear 403 error responses with explanatory payloads.

* **Chores**
  * Improved registrable-domain parsing for more reliable generic-email detection.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yashkrishan committed:
feat: Add work email validation for SSO authentication

- Block new users with generic email providers (Gmail, Yahoo, Outlook, etc.)
- Block all new GitHub signups (existing users can still sign in)
- Allow legacy users with generic emails to continue using the service
- Add server-side validation in /sso/login and /signup endpoints
- Add email domain validation utilities (extract_domain, is_generic_email)
- Return 403 Forbidden with clear error messages for blocked attempts

yashkrishan committed:
Merge 90a12df71b48079860da74fea2ced1b19a74771b into 1d33d50cca99d11fb27484a5f185be522d1afb59

github-actions[bot] committed:
chore: Auto-fix pre-commit issues

yashkrishan committed:
fix: Use publicsuffix2 for proper multi-part TLD handling

- Replace simple domain extraction with publicsuffix2.get_sld()
- Properly handles multi-part TLDs like .co.uk, .com.au
- Fixes issue where user@gmail.co.uk was incorrectly extracting to 'co.uk'
- Add fallback logic if library is not available
- Add additional Yahoo country-specific domains to blocked list

yashkrishan committed:
fix: Resolve merge conflicts and use publicsuffix2 for multi-part TLD handling

yashkrishan committed:
Merge 8240e0c279371039aa1d1d2b6f34d605070c8b14 into 1d33d50cca99d11fb27484a5f185be522d1afb59

github-actions[bot] committed:
chore: Auto-fix pre-commit issues

yashkrishan committed:
refactor: Improve extract_domain with proper email validation and module-level import

- Move publicsuffix2 import to module level with proper error handling
- Add comprehensive email validation and normalization (strip, lower)
- Improve error handling with clear fallback logic
- Add HAS_PUBLICSUFFIX2 flag to check library availability
- Ensure proper domain validation before processing
- Better documentation explaining the normalization process

yashkrishan committed:
refactor: Improve extract_domain with proper email validation and module-level import

- Move publicsuffix2 import to module level with proper error handling
- Add comprehensive email validation and normalization (strip, lower)
- Improve error handling with clear fallback logic
- Add HAS_PUBLICSUFFIX2 flag to check library availability
- Ensure proper domain validation before processing
- Better documentation explaining the normalization process

yashkrishan committed:
Merge 881effb74117f483c93dabd3b963e4aeb0c67c12 into 1d33d50cca99d11fb27484a5f185be522d1afb59

github-actions[bot] committed:
chore: Auto-fix pre-commit issues

yashkrishan committed:
fix: Remove redundant validation check that creates orphaned users

- Remove secondary validation check after authenticate_or_create
- First validation (lines 645-657) already blocks new users with generic emails
- If user reaches authenticate_or_create with generic email, they must be existing (legacy)
- Existing users return 'success' status, not 'new_user', so second check never triggers
- Removing this prevents potential orphaned user records in edge cases
- Simplifies code flow and removes unnecessary complexity

yashkrishan committed:
fix: Update publicsuffix2 to valid PyPI version

- Replace invalid version 3.0.1 with valid version 2.20191221
- Version 2.20191221 is the latest available on PyPI
- Verified availability using pip index versions

dhirenmathur commented on app/modules/auth/auth_router.py:
```diff
@@ -36,11 +36,160 @@
 auth_router = APIRouter()
 load_dotenv(override=True)
 
+# Blocked email domains (generic/personal email providers)
```
no library we can use for this instead? if not please move this to the config file

yashkrishan committed:
refactor: Move domain extraction to email_helper and use tldextract

- Replace publicsuffix2 with tldextract (more actively maintained, v5.3.0)
- Move domain extraction logic to app/modules/utils/email_helper.py
- Centralize email-related utilities in email_helper module
- Update auth_router to use extract_registrable_domain from email_helper
- Remove inline import logic from auth_router (cleaner separation of concerns)
- Update requirements.txt to use tldextract==5.3.0 instead of publicsuffix2

yashkrishan committed:
refactor: Remove duplicate email validation code, use email_helper

- Remove local BLOCKED_DOMAINS, extract_domain, and is_generic_email definitions
- Import is_personal_email_domain from app.modules.utils.email_helper
- Replace is_generic_email(email) with is_personal_email_domain(email)
- Eliminate duplicate domain lists - single source of truth in email_helper
- Cleaner code with centralized email validation utilities

yashkrishan committed:
Merge a20fcaca6bf3dd3f0408f1a0d25651b21c4646e3 into 1d33d50cca99d11fb27484a5f185be522d1afb59

github-actions[bot] committed:
chore: Auto-fix pre-commit issues
